crypto/bn/bn_gcd.c

```diff
@@ -621,3 +515,108 @@ static BIGNUM *BN_mod_inverse_no_branch(BIGNUM *in,
     bn_check_top(ret);
     return ret;
 }
+
+/*-
+ * This function is based on the constant-time GCD work by Bernstein and Yang:
+ * https://eprint.iacr.org/2019/266
+ * Generalized fast GCD function to allow even inputs.
+ * The algorithm first finds the shared powers of 2 between
+ * the inputs, and removes them, reducing at least one of the
+ * inputs to an odd value. Then it proceeds to calculate the GCD.
+ * Before returning the resulting GCD, we take care of adding
+ * back the powers of two removed at the beginning.
+ * Note: we assume the bit length of both inputs is public information,
+ * since access to top potentially leaks this information.
+ */
+int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
+{
+    BIGNUM *g, *temp = NULL;
+    BN_ULONG mask = 0;
+    int i, j, top, rlen, glen, m, bit = 1, delta = 1, cond = 0, shifts = 0, ret = 0;
+
+    /* handle corner cases immediately */
+    if (BN_is_zero(in_b)) {
+        ret = BN_copy(r, in_a) != NULL;
+        r->neg = 0;
+        return ret;
+    }
+    if (BN_is_zero(in_a)) {
+        ret = BN_copy(r, in_b) != NULL;
+        r->neg = 0;
+        return ret;
+    }
+
+    bn_check_top(in_a);
+    bn_check_top(in_b);
+
+    BN_CTX_start(ctx);
+    temp = BN_CTX_get(ctx);
+    g = BN_CTX_get(ctx);
+
+    /* make r != 0, g != 0 even, so BN_rshift is not a potential nop */
+    if (g == NULL
+        || !BN_lshift1(g, in_b)
+        || !BN_lshift1(r, in_a))
+        goto err;
+
+    /* find shared powers of two, i.e. "shifts" >= 1 */
+    for (i = 0; i < r->dmax && i < g->dmax; i++) {
+        mask = ~(r->d[i] | g->d[i]);
+        for (j = 0; j < BN_BITS2; j++) {
+            bit &= mask;
+            shifts += bit;
+            mask >>= 1;
+        }
+    }
+
+    /* subtract shared powers of two; shifts >= 1 */
+    if (!BN_rshift(r, r, shifts)
```

Comment:
In general, @cpereida did a great job of protecting what he can at this level. These `BN_rshift` and the `BN_lshift` correction at the end are examples where leakage can still occur at a lower level: if `shifts >= BN_BITS2` holds, this can be reflected in the control flow, timing, and memory access pattern of `BN_rshift`. For random inputs, this will happen with probability `2**(-2*BN_BITS2)`.

Reply:
Hmm. Taking a look at #7589, I'm a bit puzzled. `BN_lshift` was refactored as a wrapper around `bn_lshift_fixed_top`, but `BN_rshift` was not (for `bn_rshift_fixed_top`). And taking a closer look at `BN_rshift`, in combination with this PR I don't really like it at all. e.g.

https://github.com/openssl/openssl/blob/a9b5929d5686b4beef232e0f679dbae6a059f70a/crypto/bn/bn_shift.c#L199-L200

@cpereida could you have a look?

@mattcaswell it might be a separate PR but still dependent. I do agree with the code comments for `bn_lshift_fixed_top` and `bn_rshift_fixed_top`:

https://github.com/openssl/openssl/blob/a9b5929d5686b4beef232e0f679dbae6a059f70a/crypto/bn/bn_shift.c#L102-L105

AFAIK RSA is the only internal caller for `BN_gcd` and I don't feel leaking magnitude is a genuine concern there.